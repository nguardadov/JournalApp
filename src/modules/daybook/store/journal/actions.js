import journalApi from '@/api/journalApi'

//export const myAction = async({commit})=>{}
export const loadEntries = async ({ commit }) => {
  try {
    const { data } = await journalApi.get('/entries.json')

    if (!data) {
      commit('setEntries', [])
      return
    }
    const entries = []
    for (let id of Object.keys(data)) {
      entries.push({
        id,
        ...data[id],
      })
    }

    commit('setEntries', entries)
  } catch (error) {
    console.log('error: ', { error })
  }
}

export const updateEntry = async ({ commit }, entry) => {
  const { date, picture, text } = entry
  const dataTosave = { date, picture, text }

  await journalApi.put(`/entries/${entry.id}.json`, dataTosave)
  // console.log({ id, ...data })
  //commit a updateentry mutations
  dataTosave.id = entry.id
  commit('updateEntry', { ...dataTosave }) //rompiendo la referencia
}

export const createEntries = async ({ commit }, entry) => {
  const { date, picture, text } = entry
  const dataTosave = { date, picture, text }

  const { data } = await journalApi.post(`/entries.json`, dataTosave)

  dataTosave.id = data.name

  commit('addEntry', dataTosave)

  return data.name
}

export const deleteEntry = async ({ commit }, id) => {
  await journalApi.delete(`/entries/${id}.json`)
  commit('deleteEntry', id)
}
