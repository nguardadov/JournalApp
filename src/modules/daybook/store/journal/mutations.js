//export const myMutation = (state)=>{}

export const setEntries = (state, entries) => {
  state.entries = [...state.entries, ...entries] //CONVINAR ARREGLOS
  state.isLoading = false
  console.log('{entro aqui}')
}

export const updateEntry = (state, entry) => {
  // const index = state.entries.findIndex((element) => element.id === entry.id)
  // state.entries[index] = entry

  //creando un arreglo de indices
  const idx = state.entries.map((e) => e.id).indexOf(entry.id)
  state.entries[idx] = entry
}

export const addEntry = (state, entry) => {
  state.entries = [entry, ...state.entries]
}

export const deleteEntry = (state, id) => {
  // const index = state.entries.findIndex((element) => element.id === id)
  // state.entries.splice(index, 1)
  state.entries = state.entries.filter((entry) => entry.id !== id)
}

export const clearEntries = (state) => {
  state.entries = []
}
