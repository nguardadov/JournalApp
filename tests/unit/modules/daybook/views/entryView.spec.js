import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'
import Swal from 'sweetalert2'
import journal from '@/modules/daybook/store/journal'
import { journalState } from '../../../mock-data/test-journal-state'

import EntryView from '@/modules/daybook/views/EntryView.vue'

const createVuexStore = (initialState) =>
  createStore({
    modules: {
      journal: {
        ...journal, //desctructurar
        state: { ...initialState }, //aca sobreescribimos el state con lo que tengamos en initialstate
      },
    },
  })

//crear el espia
jest.mock('sweetalert2', () => ({
  fire: jest.fn(),
  showLoading: jest.fn(),
  close: jest.fn(),
}))

describe('Pruebas en el EntryView', () => {
  const store = createVuexStore(journalState)
  store.dispatch = jest.fn() //con esto simularemos la funcion del distpach

  const mockRouter = {
    push: jest.fn(),
  }
  let wrapper

  beforeEach(() => {
    jest.clearAllMocks()
    wrapper = shallowMount(EntryView, {
      props: {
        id: '-Ml2ipCYawXQcMK3BbOe',
      },
      global: {
        mocks: {
          $router: mockRouter,
        },
        plugins: [store],
      },
    })
  })

  test('debe de sacar al usuario porque el id no existe', () => {
    const wrapper = shallowMount(EntryView, {
      props: {
        id: 'este id no existe',
      },
      global: {
        mocks: {
          $router: mockRouter,
        },
        plugins: [store],
      },
    })

    expect(mockRouter.push).toHaveBeenCalledWith({ name: 'no-entry' })
  })

  test('debe de mostrar la entrada correctamente', () => {
    expect(wrapper.html()).toMatchSnapshot() //debe ser igual el snaptshop
    expect(mockRouter.push).not.toHaveBeenCalled()
  })

  test('debe de borrar la entrada y salir', (done) => {
    Swal.fire.mockReturnValueOnce(Promise.resolve({ isConfirmed: true }))
    wrapper.find('.btn-danger').trigger('click')

    expect(Swal.fire).toHaveBeenCalledWith({
      title: '¿Está seguro?',
      text: 'Una vez borrado, no se puede recuperar',
      showDenyButton: true,
      confirmButtonText: 'Si, estoy seguro',
    })

    //esto se hace para espera un milisegundo
    setTimeout(() => {
      expect(store.dispatch).toHaveBeenCalledWith(
        'journal/deleteEntry',
        '-Ml2ipCYawXQcMK3BbOe',
      )
      expect(mockRouter.push).toHaveBeenCalled()
      done()
    }, 1)
  })
})
