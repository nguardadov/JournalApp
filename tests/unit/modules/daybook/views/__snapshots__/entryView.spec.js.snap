// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">2</span><span class="mx-1 fs-3">Octubre</span><span class="mx-2 fs-4 fw-light">2021, Sábado</span></div>
  <div><input type="file" accept="image/png, image/jpeg, image/jpg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Subir Foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Qué sucedio hoy?"></textarea></div><img src="https://res.cloudinary.com/dpb6iy1ua/image/upload/v1633280934/ad9dbx9mjz3bbgo1c5rk.jpg" alt="entry-picture" class="img-thumbnail">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
